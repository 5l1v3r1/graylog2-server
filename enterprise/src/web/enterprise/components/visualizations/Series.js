import Immutable from 'immutable';

const _defaultSeriesGenerator = (type, name, labels, values) => ({ type, name, x: labels, y: values });
export const generateSeries = (config, results, chartType, generator = _defaultSeriesGenerator) => {
  const { series } = config;
  const rowPivots = config.rowPivots.map(({ field }) => field);
  const columnPivots = config.columnPivots.map(({ field }) => field);
  const x = results.map(v => rowPivots.map(p => v[p]).join('-'));

  const columnPivotCharts = columnPivots.map((columnPivot) => {
    const values = {};
    const columnPivotX = results.filter(r => r[columnPivot])
      .map(row => row[columnPivot].map(pivot => pivot[columnPivot]))
      .reduce((prev, cur) => prev.merge(cur), Immutable.OrderedSet()).toJS();
    results.forEach((row) => {
      const valueKey = rowPivots.map(p => row[p]).join('-');
      columnPivotX.forEach((columnPivotValue) => {
        series.forEach((seriesName) => {
          const pivotValueKey = `${columnPivotValue}-${seriesName}`;

          if (!values[pivotValueKey]) {
            values[pivotValueKey] = [];
          }

          const rowValue = (row[columnPivot].find(r => r[columnPivot] === columnPivotValue) || {})[seriesName];
          values[`${columnPivotValue}-${seriesName}`].push({ key: valueKey, value: rowValue });
        });
      });
    });

    return Object.keys(values).map(pivotValue => [
      chartType,
      pivotValue,
      values[pivotValue].map(({ key }) => key),
      values[pivotValue].map(({ value }) => value),
    ]);
  }).flatten();

  const seriesCharts = series.map(s => [chartType, s, x, results.map(v => v[s])]);

  const allCharts = [].concat(seriesCharts, columnPivotCharts);
  return allCharts.map((args, idx) => generator(...args, idx, allCharts.length));
};
