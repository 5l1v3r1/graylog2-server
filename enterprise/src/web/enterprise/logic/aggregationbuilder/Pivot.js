export default class Pivot {
  constructor(field, type, config = {}) {
    this._value = { field, type, config };
  }
  get field() {
    return this._value.field;
  }

  get type() {
    return this._value.type;
  }

  get config() {
    return this._value.config;
  }

  static fromJSON(value) {
    const { field, type, config = {} } = value;
    return new Pivot(field, type, config);
  }

  toJSON() {
    const { field, type, config } = this._value;
    return { field, type, config };
  }
}
