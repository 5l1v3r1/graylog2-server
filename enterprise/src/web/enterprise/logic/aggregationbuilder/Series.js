// @flow strict
import { get } from 'lodash';
import * as Immutable from 'immutable';

import SeriesConfig from './SeriesConfig';
import type { SeriesConfigJson } from './SeriesConfig';

export type SeriesJson = {
  config: SeriesConfigJson,
  function: string,
};

type InternalState = {
  config: SeriesConfig,
  function: string,
};

export default class Series {
  _value: InternalState;
  constructor(func: string, config: SeriesConfig = SeriesConfig.empty()) {
    this._value = { function: func, config };
  }

  get function() {
    return this._value.function;
  }

  get config() {
    return this._value.config;
  }

  get effectiveName() {
    const overridenName = get(this, 'config.name');
    return overridenName || this.function;
  }

  toString() {
    return `Series: ${this.effectiveName}, config={${JSON.stringify(this.config)}}`;
  }

  toJSON() {
    return {
      config: this._value.config,
      function: this._value.function,
    };
  }

  static fromJSON(value: SeriesJson) {
    return new Series(value.function, SeriesConfig.fromJSON(value.config));
  }

  static forFunction(func: string) {
    // eslint-disable-next-line no-use-before-define
    return new Builder().function(func).config(SeriesConfig.empty()).build();
  }

  toBuilder() {
    // eslint-disable-next-line no-use-before-define
    return new Builder(Immutable.Map(this._value));
  }
}

type BuilderState = Immutable.Map<string, any>;
class Builder {
  value: BuilderState;
  constructor(value: BuilderState = Immutable.Map()) {
    this.value = value;
  }

  function(newFunction: string) {
    return new Builder(this.value.set('function', newFunction));
  }

  config(newConfig: SeriesConfig) {
    return new Builder(this.value.set('config', newConfig));
  }

  build() {
    const { config } = this.value.toObject();
    const func = this.value.get('function');
    return new Series(func, config);
  }
}
