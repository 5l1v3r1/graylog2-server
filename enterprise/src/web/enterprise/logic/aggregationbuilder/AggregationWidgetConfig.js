// @flow strict
import * as Immutable from 'immutable';

import { TIMESTAMP_FIELD } from 'enterprise/Constants';
import Pivot from './Pivot';
import Series from './Series';
import VisualizationConfig from './visualizations/VisualizationConfig';
import SortConfig from './SortConfig';
import type { SeriesJson } from './Series';
import type { SortConfigJson } from './SortConfig';
import type { VisualizationConfigJson } from './visualizations/VisualizationConfig';
import type { PivotJson } from './Pivot';

type InternalState = {
  columnPivots: Array<Pivot>,
  rowPivots: Array<Pivot>,
  series: Array<Series>,
  sort: Array<SortConfig>,
  visualization: string,
  rollup: boolean,
  visualizationConfig: ?VisualizationConfig,
};

type AggregationWidgetConfigJson = {
  row_pivots: Array<PivotJson>,
  column_pivots: Array<PivotJson>,
  series: Array<SeriesJson>,
  sort: Array<SortConfigJson>,
  visualization: string,
  rollup: boolean,
  visualization_config: VisualizationConfigJson,
};

export default class AggregationWidgetConfig {
  _value: InternalState;

  constructor(columnPivots: Array<Pivot>,
    rowPivots: Array<Pivot>,
    series: Array<Series>,
    sort: Array<SortConfig>,
    visualization: string,
    rollup: boolean,
    visualizationConfig: VisualizationConfig) {
    this._value = { columnPivots, rowPivots, series, sort, visualization, rollup, visualizationConfig };
  }

  get rowPivots() {
    return this._value.rowPivots;
  }

  get columnPivots() {
    return this._value.columnPivots;
  }

  get series() {
    return this._value.series;
  }

  get sort() {
    return this._value.sort;
  }

  get rollup() {
    return this._value.rollup;
  }

  get visualization() {
    return this._value.visualization;
  }

  get visualizationConfig() {
    return this._value.visualizationConfig;
  }

  get isTimeline() {
    return this.rowPivots && this.rowPivots.length === 1 && this.rowPivots[0].field === TIMESTAMP_FIELD;
  }

  static builder() {
    // eslint-disable-next-line no-use-before-define
    return new Builder()
      .rowPivots([])
      .columnPivots([])
      .series([])
      .sort([])
      .rollup(true);
  }

  toBuilder() {
    // eslint-disable-next-line no-use-before-define
    return new Builder(Immutable.Map(this._value));
  }

  toJSON() {
    const { rowPivots, columnPivots, series, sort, visualization, visualizationConfig, rollup } = this._value;
    return {
      row_pivots: rowPivots,
      column_pivots: columnPivots,
      series,
      sort,
      visualization,
      visualization_config: visualizationConfig,
      rollup,
    };
  }

  equals(other: any) {
    const { is } = Immutable;
    if (other instanceof AggregationWidgetConfig) {
      return ['rowPivots', 'columnPivots', 'series', 'sort', 'rollup']
        // $FlowFixMe: No typings for indexed access.
        .every(key => is(this[key], other[key]));
    }
    return false;
  }

  static fromJSON(value: AggregationWidgetConfigJson) {
    // eslint-disable-next-line camelcase
    const { row_pivots, column_pivots, series, sort, visualization, rollup, visualization_config } = value;

    // eslint-disable-next-line no-use-before-define
    return new Builder()
      .columnPivots(column_pivots.map(Pivot.fromJSON))
      .rowPivots(row_pivots.map(Pivot.fromJSON))
      .series(series.map(Series.fromJSON))
      .sort(sort.map(SortConfig.fromJSON))
      .visualization(visualization)
      .rollup(rollup)
      // eslint-disable-next-line camelcase
      .visualizationConfig(visualization_config !== null ? VisualizationConfig.fromJSON(visualization, visualization_config) : null)
      .build();
  }
}

type BuilderState = Immutable.Map<string, any>;
class Builder {
  value: BuilderState;
  constructor(value: BuilderState = Immutable.Map()) {
    this.value = value;
  }

  columnPivots(pivots: Array<Pivot>) {
    return new Builder(this.value.set('columnPivots', pivots));
  }

  rowPivots(pivots: Array<Pivot>) {
    return new Builder(this.value.set('rowPivots', pivots));
  }

  series(series: Array<Series>) {
    return new Builder(this.value.set('series', series));
  }

  sort(sorts: Array<SortConfig>) {
    return new Builder(this.value.set('sort', sorts));
  }

  visualization(type: string) {
    return new Builder(this.value.set('visualization', type));
  }

  visualizationConfig(config: ?VisualizationConfig) {
    return new Builder(this.value.set('visualizationConfig', config));
  }

  rollup(rollup: boolean) {
    return new Builder(this.value.set('rollup', rollup));
  }

  build() {
    const { rowPivots, columnPivots, series, sort, visualization, rollup, visualizationConfig } = this.value.toObject();

    const availableSorts = [].concat(rowPivots, columnPivots, series);
    const filteredSorts = sort.filter(s => availableSorts.find(availableSort => (s.field === availableSort.function || s.field === availableSort.field)));
    const computedRollup = columnPivots.length > 0 ? rollup : true;
    return new AggregationWidgetConfig(columnPivots, rowPivots, series, filteredSorts, visualization, computedRollup, visualizationConfig);
  }
}
