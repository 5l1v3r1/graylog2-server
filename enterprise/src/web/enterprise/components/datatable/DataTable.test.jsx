// @flow strict
import React from 'react';
import renderer from 'react-test-renderer';
import Immutable from 'immutable';
// $FlowFixMe: imports from core need to be fixed in flow
import { CombinedProviderMock, StoreMock } from 'helpers/mocking';
import AggregationWidgetConfig from 'enterprise/logic/aggregationbuilder/AggregationWidgetConfig';
import Pivot from 'enterprise/logic/aggregationbuilder/Pivot';
import Series from 'enterprise/logic/aggregationbuilder/Series';
// $FlowFixMe: imports from core need to be fixed in flow
import 'helpers/mocking/react-dom_mock';

describe('DataTable', () => {
  const CurrentUserStore = StoreMock('listen', 'get');
  const combinedProviderMock = new CombinedProviderMock({
    CurrentUser: CurrentUserStore,
  });

  jest.doMock('injection/CombinedProvider', () => combinedProviderMock);

  /* eslint-disable-next-line global-require */
  const DataTable = require('./DataTable').default;

  const currentView = { activeQuery: 'deadbeef-23' };

  const data = [{
    key: ['2018-10-04T09:43:50.000Z'],
    source: 'leaf',
    values: [{
      key: ['hulud.net', 'count()'],
      rollup: false,
      source: 'col-leaf',
      value: 408,
    }, {
      key: ['count()'],
      rollup: true,
      source: 'row-leaf',
      value: 408,
    }],
  }];

  const columnPivot = new Pivot('source', 'values', { limit: 15 });
  const rowPivot = new Pivot('timestamp', 'time', { interval: 'auto' });
  const series = new Series('count()');

  it('should render with empty data', () => {
    const config = AggregationWidgetConfig.builder()
      .rowPivots([])
      .columnPivots([])
      .series([])
      .sort([])
      .visualization('table')
      .rollup(true)
      .build();
    const wrapper = renderer.create(<DataTable config={config}
                                               currentView={currentView}
                                               data={[]}
                                               fields={Immutable.List([])} />);
    expect(wrapper.toJSON()).toMatchSnapshot();
  });

  it('should render with filled data with rollup', () => {
    const config = AggregationWidgetConfig.builder()
      .rowPivots([rowPivot])
      .columnPivots([columnPivot])
      .series([series])
      .sort([])
      .visualization('table')
      .rollup(true)
      .build();
    const wrapper = renderer.create(<DataTable config={config}
                                               currentView={currentView}
                                               data={data}
                                               fields={Immutable.List([])} />);
    expect(wrapper.toJSON()).toMatchSnapshot();
  });

  it('should render with filled data without rollup', () => {
    const config = AggregationWidgetConfig.builder()
      .rowPivots([rowPivot])
      .columnPivots([columnPivot])
      .series([series])
      .sort([])
      .visualization('table')
      .rollup(false)
      .build();
    const wrapper = renderer.create(<DataTable config={config}
                                               currentView={currentView}
                                               data={data}
                                               fields={Immutable.List([])} />);
    expect(wrapper.toJSON()).toMatchSnapshot();
  });

  it('renders column pivot header without offset when rollup is disabled', () => {
    const protocolPivot = new Pivot('nf_proto_name', 'values', { limit: 15 });
    const protocolData = [{
      key: [],
      values: [{
        key: ['TCP', 'count()'],
        value: 239,
        rollup: false,
        source: 'col-leaf',
      }, { key: ['UDP', 'count()'], value: 226, rollup: false, source: 'col-leaf' }],
      source: 'leaf',
    }];

    const config = AggregationWidgetConfig.builder()
      .rowPivots([])
      .columnPivots([protocolPivot])
      .series([series])
      .sort([])
      .visualization('table')
      .rollup(false)
      .build();
    const wrapper = renderer.create(<DataTable config={config}
                                               currentView={currentView}
                                               data={protocolData}
                                               fields={Immutable.List([])} />);
    expect(wrapper.toJSON()).toMatchSnapshot();
  });
});
