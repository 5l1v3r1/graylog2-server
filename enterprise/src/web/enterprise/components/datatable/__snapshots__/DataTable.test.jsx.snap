// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DataTable renders column pivot header without offset when rollup is disabled 1`] = `
<div
  className="messages-container"
  style={
    Object {
      "height": "100%",
      "overflow": "auto",
    }
  }
>
  <table
    className="table table-condensed messages"
  >
    <thead>
      <tr>
        <th
          colSpan={1}
          style={
            Object {
              "left": "0px",
            }
          }
        >
          <span>
            <span
              className="dropdowntoggle"
              onClick={[Function]}
              role="presentation"
            >
              TCP
              <span
                className="caret"
              />
            </span>
          </span>
        </th>
        <th
          colSpan={1}
          style={
            Object {
              "left": "0px",
            }
          }
        >
          <span>
            <span
              className="dropdowntoggle"
              onClick={[Function]}
              role="presentation"
            >
              UDP
              <span
                className="caret"
              />
            </span>
          </span>
        </th>
      </tr>
      <tr>
        <th
          colSpan={1}
          style={
            Object {
              "left": "0px",
            }
          }
        >
          <span>
            <span
              className="dropdowntoggle"
              onClick={[Function]}
              role="presentation"
            >
              <span
                className="field-element  "
              >
                count()
              </span>
              <span
                className="caret"
              />
            </span>
          </span>
        </th>
        <th
          colSpan={1}
          style={
            Object {
              "left": "0px",
            }
          }
        >
          <span>
            <span
              className="dropdowntoggle"
              onClick={[Function]}
              role="presentation"
            >
              <span
                className="field-element  "
              >
                count()
              </span>
              <span
                className="caret"
              />
            </span>
          </span>
        </th>
      </tr>
    </thead>
    <tbody
      className="message-group"
    >
      <tr
        className="fields-row"
      >
        <td>
          <span>
            <span
              className="dropdowntoggle"
              onClick={[Function]}
              role="presentation"
            >
              239
              <span
                className="caret"
              />
            </span>
          </span>
        </td>
        <td>
          <span>
            <span
              className="dropdowntoggle"
              onClick={[Function]}
              role="presentation"
            >
              226
              <span
                className="caret"
              />
            </span>
          </span>
        </td>
      </tr>
    </tbody>
  </table>
</div>
`;

exports[`DataTable should render with empty data 1`] = `
<div
  className="messages-container"
  style={
    Object {
      "height": "100%",
      "overflow": "auto",
    }
  }
>
  <table
    className="table table-condensed messages"
  >
    <thead>
      <tr />
    </thead>
  </table>
</div>
`;

exports[`DataTable should render with filled data with rollup 1`] = `
<div
  className="messages-container"
  style={
    Object {
      "height": "100%",
      "overflow": "auto",
    }
  }
>
  <table
    className="table table-condensed messages"
  >
    <thead>
      <tr>
        <th
          colSpan={2}
          style={
            Object {
              "left": "0px",
            }
          }
        />
        <th
          colSpan={1}
          style={
            Object {
              "left": "0px",
            }
          }
        >
          <span>
            <span
              className="dropdowntoggle"
              onClick={[Function]}
              role="presentation"
            >
              hulud.net
              <span
                className="caret"
              />
            </span>
          </span>
        </th>
      </tr>
      <tr>
        <th
          colSpan={
            Array [
              "timestamp",
            ]
          }
          style={
            Object {
              "left": "0px",
            }
          }
        >
          <span>
            <span
              className="dropdowntoggle"
              onClick={[Function]}
              role="presentation"
            >
              <span
                className="field-element  "
              >
                timestamp
              </span>
              <span
                className="caret"
              />
            </span>
          </span>
        </th>
        <th
          colSpan={
            Array [
              "count()",
            ]
          }
          style={
            Object {
              "left": "0px",
            }
          }
        >
          <span>
            <span
              className="dropdowntoggle"
              onClick={[Function]}
              role="presentation"
            >
              <span
                className="field-element  "
              >
                count()
              </span>
              <span
                className="caret"
              />
            </span>
          </span>
        </th>
        <th
          colSpan={1}
          style={
            Object {
              "left": "0px",
            }
          }
        >
          <span>
            <span
              className="dropdowntoggle"
              onClick={[Function]}
              role="presentation"
            >
              <span
                className="field-element  "
              >
                count()
              </span>
              <span
                className="caret"
              />
            </span>
          </span>
        </th>
      </tr>
    </thead>
    <tbody
      className="message-group"
    >
      <tr
        className="fields-row"
      >
        <td>
          <span>
            <span
              className="dropdowntoggle"
              onClick={[Function]}
              role="presentation"
            >
              2018-10-04T09:43:50.000Z
              <span
                className="caret"
              />
            </span>
          </span>
        </td>
        <td>
          <span>
            <span
              className="dropdowntoggle"
              onClick={[Function]}
              role="presentation"
            >
              408
              <span
                className="caret"
              />
            </span>
          </span>
        </td>
        <td>
          <span>
            <span
              className="dropdowntoggle"
              onClick={[Function]}
              role="presentation"
            >
              408
              <span
                className="caret"
              />
            </span>
          </span>
        </td>
      </tr>
    </tbody>
  </table>
</div>
`;

exports[`DataTable should render with filled data without rollup 1`] = `
<div
  className="messages-container"
  style={
    Object {
      "height": "100%",
      "overflow": "auto",
    }
  }
>
  <table
    className="table table-condensed messages"
  >
    <thead>
      <tr>
        <th
          colSpan={1}
          style={
            Object {
              "left": "0px",
            }
          }
        />
        <th
          colSpan={1}
          style={
            Object {
              "left": "0px",
            }
          }
        >
          <span>
            <span
              className="dropdowntoggle"
              onClick={[Function]}
              role="presentation"
            >
              hulud.net
              <span
                className="caret"
              />
            </span>
          </span>
        </th>
      </tr>
      <tr>
        <th
          colSpan={
            Array [
              "timestamp",
            ]
          }
          style={
            Object {
              "left": "0px",
            }
          }
        >
          <span>
            <span
              className="dropdowntoggle"
              onClick={[Function]}
              role="presentation"
            >
              <span
                className="field-element  "
              >
                timestamp
              </span>
              <span
                className="caret"
              />
            </span>
          </span>
        </th>
        <th
          colSpan={1}
          style={
            Object {
              "left": "0px",
            }
          }
        >
          <span>
            <span
              className="dropdowntoggle"
              onClick={[Function]}
              role="presentation"
            >
              <span
                className="field-element  "
              >
                count()
              </span>
              <span
                className="caret"
              />
            </span>
          </span>
        </th>
      </tr>
    </thead>
    <tbody
      className="message-group"
    >
      <tr
        className="fields-row"
      >
        <td>
          <span>
            <span
              className="dropdowntoggle"
              onClick={[Function]}
              role="presentation"
            >
              2018-10-04T09:43:50.000Z
              <span
                className="caret"
              />
            </span>
          </span>
        </td>
        <td>
          <span>
            <span
              className="dropdowntoggle"
              onClick={[Function]}
              role="presentation"
            >
              408
              <span
                className="caret"
              />
            </span>
          </span>
        </td>
      </tr>
    </tbody>
  </table>
</div>
`;
